src/main/java/de/plushnikov/intellij/plugin/hack/extension/EMSupport.java

```diff
@@ -0,0 +1,107 @@
+import com.intellij.psi.scope.PsiScopeProcessor;
+
+import de.plushnikov.intellij.plugin.hack.Injector;
+import org.jetbrains.org.objectweb.asm.ClassReader;
+import org.jetbrains.org.objectweb.asm.ClassWriter;
+import org.jetbrains.org.objectweb.asm.tree.AbstractInsnNode;
+import org.jetbrains.org.objectweb.asm.tree.ClassNode;
+import org.jetbrains.org.objectweb.asm.tree.FieldInsnNode;
+import org.jetbrains.org.objectweb.asm.tree.InsnList;
+import org.jetbrains.org.objectweb.asm.tree.InsnNode;
+import org.jetbrains.org.objectweb.asm.tree.MethodInsnNode;
+import org.jetbrains.org.objectweb.asm.tree.MethodNode;
+import org.jetbrains.org.objectweb.asm.tree.VarInsnNode;
+
+import static org.jetbrains.org.objectweb.asm.Opcodes.*;
+
+public class EMSupport {
+
+  static {
+    try {
+      final ClassLoader loader = PsiClassImplUtil.class.getClassLoader();
+      final Method defineClass = ClassLoader.class.getDeclaredMethod("defineClass", byte[].class, int.class, int.class);
+      defineClass.setAccessible(true);
+      final byte[] bytes = getBytesFromClass(EMHolder.class);
+      final Class<?> holder = (Class<?>) defineClass.invoke(loader, bytes, 0, bytes.length);
+      final Field handle = holder.getField("handle");
+      handle.setAccessible(true);
+      handle.set(null, MethodHandles.lookup().findStatic(EMHandler.class, "processDeclarations",
+        MethodType.methodType(boolean.class, boolean.class, PsiClass.class, PsiScopeProcessor.class, ResolveState.class, PsiElement.class, PsiElement.class)));
+      final Instrumentation instrumentation = Injector.instrumentation();
+      instrumentation.addTransformer(new ClassFileTransformer() {
+        @Override
+        public byte[] transform(final ClassLoader loader, final String className, final Class<?> classBeingRedefined, final ProtectionDomain protectionDomain, final byte[] bytes) throws IllegalClassFormatException {
+          if (classBeingRedefined == PsiClassImplUtil.class) {
+            final ClassNode node = new ClassNode();
+            final ClassReader reader = new ClassReader(bytes);
+            reader.accept(node, 0);
+            for (final MethodNode method : node.methods) {
+              if (method.name.equals("processDeclarationsInClass") && method.desc.equals("(Lcom/intellij/psi/PsiClass;Lcom/intellij/psi/scope/PsiScopeProcessor;Lcom/intellij/psi/ResolveState;Ljava/util/Set;Lcom/intellij/psi/PsiElement;Lcom/intellij/psi/PsiElement;Lcom/intellij/pom/java/LanguageLevel;ZLcom/intellij/psi/search/GlobalSearchScope;)Z")) {
+                final InsnList instructions = method.instructions;
+                int count = 2;
+                final List<AbstractInsnNode> target = new ArrayList<>();
+                for (final ListIterator<AbstractInsnNode> iterator = instructions.iterator(instructions.size()); iterator.hasPrevious() && count > 0; ) {
+                  final AbstractInsnNode insn = iterator.previous();
+                  if (insn.getOpcode() == IRETURN) {
```

Comment:
Ok, as I see, you inject yourself before every return statement. What do you think would be a best place to add normal ExtensionPoint to this method? If you want, provide a patch for this and I'l look how to apply it to intelliJ sources to remove need for bytecode instrumentation.

Reply:
Putting the extension point in PsiClassImplUtil#processDeclarationsInClass is the simplest solution, just pass the arguments of this method to the corresponding method of the extension point as well.